ee/clickhouse/process_event.py

```diff
@@ -0,0 +1,181 @@
+import datetime
+from typing import Dict, List, Optional, Union
+
+from celery import shared_task
+from django.db import IntegrityError
+
+from ee.clickhouse.models.element import create_elements
+from ee.clickhouse.models.event import create_event
+from ee.clickhouse.models.person import attach_distinct_ids, create_person_with_distinct_id
+from posthog.models.element import Element
+from posthog.models.person import Person
+from posthog.models.team import Team
+from posthog.tasks.process_event import check_and_create_person, handle_timestamp, store_names_and_properties
+
+
+def _alias(previous_distinct_id: str, distinct_id: str, team_id: int, retry_if_failed: bool = True,) -> None:
+    old_person: Optional[Person] = None
+    new_person: Optional[Person] = None
+
+    try:
+        old_person = Person.objects.get(team_id=team_id, persondistinctid__distinct_id=previous_distinct_id)
+    except Person.DoesNotExist:
+        pass
+
+    try:
+        new_person = Person.objects.get(team_id=team_id, persondistinctid__distinct_id=distinct_id)
+    except Person.DoesNotExist:
+        pass
+
+    if old_person and not new_person:
+        try:
+            old_person.add_distinct_id(distinct_id)
+            attach_distinct_ids(old_person.pk, [distinct_id], team_id)
+        # Catch race case when somebody already added this distinct_id between .get and .add_distinct_id
+        except IntegrityError:
+            if retry_if_failed:  # run everything again to merge the users if needed
+                _alias(previous_distinct_id, distinct_id, team_id, False)
+        return
+
+    if not old_person and new_person:
+        try:
+            new_person.add_distinct_id(previous_distinct_id)
+            attach_distinct_ids(new_person.pk, [previous_distinct_id], team_id)
+        # Catch race case when somebody already added this distinct_id between .get and .add_distinct_id
+        except IntegrityError:
+            if retry_if_failed:  # run everything again to merge the users if needed
+                _alias(previous_distinct_id, distinct_id, team_id, False)
+        return
+
+    if not old_person and not new_person:
+        try:
+            person = Person.objects.create(team_id=team_id, distinct_ids=[str(distinct_id), str(previous_distinct_id)],)
+            create_person_with_distinct_id(
+                person_id=person.pk, team_id=team_id, distinct_ids=[str(distinct_id), str(previous_distinct_id)],
+            )
+        # Catch race condition where in between getting and creating, another request already created this user.
+        except IntegrityError:
+            if retry_if_failed:
+                # try once more, probably one of the two persons exists now
+                _alias(previous_distinct_id, distinct_id, team_id, False)
+        return
+
+    if old_person and new_person and old_person != new_person:
+        new_person.merge_people([old_person])
```

Comment:
Can't adapt this function without implementing incremental change structure in clickhouse